Implement CMake build system

Timmmm opened the pull request:
Switch from Makefiles to CMake. Some formal outputs are not yet fully supported by the CMake so the old Makefile is retained.
    
CMake has significant advantages:

* IDE support is perfect, since it generates `compile_commands.json`. You can even go-to-definition to the generated code (I can demo this in the meeting if people want). Lots of IDEs also understand CMake and CTest.
* Building all the emulators and running the tests is easy, parallel, and can generate an XML test result file (I haven't tested this though).
* Better Windows support, e.g. CMake can generate Visual Studio project files.

This also adds a manual Github actions workflow to generate the JSON doc bundle, to allow inclusion of Sail snippets in the ISA manual.

arichardson commented on emulator/riscv_sim_RV64:
binary added here

arichardson commented on CMakeLists.txt:
```diff
@@ -0,0 +1,76 @@
+cmake_minimum_required(VERSION 3.27)
```
That is quite a new version, are any of the new features needed? If not it would be good to use something older that is available on most distros.

Timmmm replied:
Good point. I'll reduce it to 3.22 which is the version in Ubuntu 22.04.

Btw an easy way to install the latest version of CMake is `python3 -m pip install cmake`. It's officially maintained and has the latest version. The only snag with this technique is that Debian has started requiring `--break-system-packages` (it's perfectly safe in this case).

arichardson commented on CMakeLists.txt:
```diff
@@ -0,0 +1,76 @@
+    message(FATAL_ERROR "GMP not found. Try 'sudo apt install libgmp3-dev' or 'sudo dnf install gmp-devel'.")
+endif()
+
+find_program(sail "sail")
+if (NOT sail)
```
```suggestion
find_program(SAIL_BIN "sail")
if (NOT SAIL_BIN)
```

arichardson commented on cmake/modules/FindGMP.cmake:
```diff
@@ -0,0 +1,51 @@
+#  GMP_C_LIBRARIES - LIST:The libraries needed to use GMP via it's C interface
+#  GMP_CXX_LIBRARIES - LIST:The libraries needed to use GMP via it's C++ interface
+
+include(FindPackageHandleStandardArgs)
+
+# Try to find libraries
+find_library(GMP_C_LIBRARIES
+  NAMES gmp
+  DOC "GMP C libraries"
+)
+find_library(GMP_CXX_LIBRARIES
+  NAMES gmpxx
+  DOC "GMP C++ libraries"
+)
+
+# Try to find headers
+find_path(GMP_C_INCLUDES
+  NAMES gmp.h
+  DOC "GMP C header"
+)
+
+find_path(GMP_CXX_INCLUDES
+  NAMES gmpxx.h
+  DOC "GMP C++ header"
+)
```
```suggestion
include(FindPackageHandleStandardArgs)
find_package(PkgConfig)
if(PKG_CONFIG_FOUND)
  pkg_check_modules(PC_GMP QUIET gmp)
  pkg_check_modules(PC_GMPXX QUIET gmpxx)
endif()

# Try to find libraries
find_library(GMP_C_LIBRARIES
  NAMES gmp
  PATHS ${PC_GMP_LIBRARY_DIRS}
  DOC "GMP C libraries"
)
find_library(GMP_CXX_LIBRARIES
  NAMES gmpxx
  PATHS ${PC_GMPXX_LIBRARY_DIRS}
  DOC "GMP C++ libraries"
)

# Try to find headers
find_path(GMP_C_INCLUDES
  NAMES gmp.h
  PATHS ${PC_GMP_INCLUDE_DIRS}
  DOC "GMP C header"
)

find_path(GMP_CXX_INCLUDES
  NAMES gmpxx.h
  PATHS ${PC_GMPXX_INCLUDE_DIRS}
  DOC "GMP C++ header"
)
```

Following https://cmake.org/cmake/help/latest/manual/cmake-developer.7.html#a-sample-find-module

arichardson commented on cmake/modules/FindGMP.cmake:
```diff
@@ -0,0 +1,51 @@
+# Handle QUIET and REQUIRED and check the necessary variables were set and if so
+# set ``GMP_FOUND``
+find_package_handle_standard_args(GMP
+	REQUIRED_VARS GMP_C_LIBRARIES GMP_C_INCLUDES GMP_CXX_LIBRARIES GMP_CXX_INCLUDES)
```
Since GMP++ is not used, maybe drop that or add it as a separate find module?

Timmmm replied:
Good point, I'll drop it.

arichardson commented on emulator/CMakeLists.txt:
```diff
@@ -0,0 +1,41 @@
+foreach (arch IN LISTS ENABLED_ARCHITECTURES)
+
+    add_executable(emulator_${arch}
+        "${CMAKE_BINARY_DIR}/riscv_model_${arch}.c"
+        "riscv_config.h"
+        "riscv_platform.c"
+        "riscv_platform.h"
+        "riscv_platform_impl.c"
+        "riscv_platform_impl.h"
+        "riscv_prelude.c"
+        "riscv_prelude.h"
+        "riscv_sail.h"
+        "riscv_sim.c"
+        "riscv_softfloat.c"
+        "riscv_softfloat.h"
+    )
```
```suggestion
set(EMULATOR_COMMON_SRCS 
        riscv_config.h
        riscv_platform.c
        riscv_platform.h
        riscv_platform_impl.c
        riscv_platform_impl.h
        riscv_prelude.c
        riscv_prelude.h
        riscv_sail.h
        riscv_sim.c
        riscv_softfloat.c
        riscv_softfloat.h
)
foreach (arch IN LISTS ENABLED_ARCHITECTURES)

    add_executable(emulator_${arch}
        ${CMAKE_BINARY_DIR}/riscv_model_${arch}.c
        ${EMULATOR_COMMON_SRCS}
    )
```

arichardson commented on emulator/CMakeLists.txt:
```diff
@@ -0,0 +1,41 @@
+    add_dependencies(emulator_${arch} generated_model_${arch})
+
+    target_link_libraries(emulator_${arch}
+        PRIVATE softfloat sail_runtime gmp z
```
```suggestion
        PRIVATE softfloat sail_runtime GMP::GMP ZLIB::ZLIB
```
Use the imported targets from the find modules?

arichardson reviewed (commented):
Overall this is looking good!
I've left a few CMake comments, but it's quite hard to review with the softfloat upgrade included here.

Timmmm commented:
Thanks for the review! I've split the other changes out into separate commits, so the CMake stuff is just the last commit now and hopefully easier to review.

I might drop the upgrade to softfloat actually since it didn't really change the code and hopefully we'll remove it at some point anyway in favour of the pure Sail implementation.

jordancarlin commented:
I think it probably makes sense to skip the softfloat upgrade for the reasons you mentioned, but if we do decide to keep it I have some suggestions so that you don’t have to modify their Makefile.

arichardson reviewed (commented):
Regarding the commit that suggests bundling GMP - we could also use https://cmake.org/cmake/help/latest/module/ExternalProject.html#examples to build a known version statically (either opt-in or by default).

arichardson commented on Makefile:
```diff
@@ -1,428 +0,0 @@
-# Select architecture: RV32 or RV64.
```
Could keep the makefile for the targets that aren't supported yet and just defer to CMake by default?
e.g.
```
csim: $(BUILD_DIR)
    cmake -S . -B $(BUILD_DIR) -GNinja && ninja -C $(BUILD_DIR)
```

Or to avoid needing ninja:
```
csim: $(BUILD_DIR)
    cmake -S . -B $(BUILD_DIR) && $(MAKE) -C $(BUILD_DIR)
```

arichardson commented on .gitignore:
```diff
@@ -6,4 +6,7 @@ _build/
+z3_problems
+
+# Typical CMake build directory.
+/build
```
Could you also add `/cmake-build-*` which is the default for CLion (it creates cmake-build-debug/cmake-build-release, etc. depending on the CMake mode you choose.

Timmmm replied:
Added

arichardson commented on CMakeLists.txt:
```diff
@@ -0,0 +1,76 @@
+cmake_minimum_required(VERSION 3.22)
```
As noted before this seems quite high, can we use a lower min version?

arichardson replied:
LLVM requires 3.20, and check_linker_flag was added in 3.18, so maybe 3.20 is a good base version?

Timmmm replied:
Changed to 3.20

arichardson commented on CMakeLists.txt:
```diff
@@ -0,0 +1,76 @@
+# much older compilers support this feature so everything does work with C++17,
+# but you get lots of warnings on compilers that support C++20 if you use
+# this feature without -std=c++20.
+if (cxx_std_20 IN_LIST CMAKE_CXX_COMPILE_FEATURES)
+    set(CMAKE_CXX_STANDARD 20)
+else()
+    set(CMAKE_CXX_STANDARD 17)
+endif()
+set(CMAKE_CXX_STANDARD_REQUIRED TRUE)
```
I think it should also be possible to use the default fallback to older version behaviour:
```
set(CMAKE_CXX_STANDARD 20)
if (cxx_std_20 IN_LIST CMAKE_CXX_COMPILE_FEATURES)
    set(CMAKE_CXX_STANDARD_REQUIRED TRUE)
endif()
```
But as far as I can tell the current sail code does not rely on this and the files are compiled as C?

Timmmm replied:
Ah good point - this was needed for the `struct Model { #include "model.c" }` trick which isn't used yet. I'll just remove it for now since there isn't any C++ yet.

arichardson commented on CMakeLists.txt:
```diff
@@ -0,0 +1,76 @@
+set(CMAKE_POSITION_INDEPENDENT_CODE TRUE)
+
+# Don't allow undefined symbols in shared libraries. This is generally a pain.
+if (UNIX)
```
I think it would be cleaner to use https://cmake.org/cmake/help/latest/module/CheckLinkerFlag.html

```
check_linker_flag(C -Wl,--no-undefined LINKER_SUPPORTS_NO_UNDEFINED)
if (LINKER_SUPPORTS_NO_UNDEFINED)
    add_link_options("-Wl,--no-undefined")
endif()

```

Timmmm replied:
Good point, done.

arichardson commented on CMakeLists.txt:
```diff
@@ -0,0 +1,76 @@
+set(CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake/modules")
+
+# These are the main requirements.
+# Don't use `REQUIRED` so that we can print custom help messages.
```
CMake has support for custom messages with https://cmake.org/cmake/help/latest/module/FeatureSummary.html
and the set_package_properties() helper.

Timmmm replied:
Huh TIL! Unfortunately it doesn't exactly support a "message if not found", and it also doesn't work with `find_program`... best I could come up with is this which I'm not sure is really better:

```
find_package(ZLIB)
find_package(GMP)
find_program(SAIL_BIN "sail")

include(FeatureSummary)

set_package_properties(ZLIB PROPERTIES
    URL "https://zlib.net/"
    DESCRIPTION "Zlib compression library"
    TYPE REQUIRED
    PURPOSE "Allows reading .elf.gz files.\n   Install with 'sudo apt install zlib1g-dev' or 'sudo dnf install zlib-devel'."
)

set_package_properties(GMP PROPERTIES
    URL "https://gmplib.org/"
    DESCRIPTION "GNU Multi Precision library"
    TYPE REQUIRED
    PURPOSE "Sail C code requires this to support arbitrary sized integers.\n   Install with 'sudo apt install libgmp3-dev' or 'sudo dnf install gmp-devel'."
)

feature_summary(
    FATAL_ON_MISSING_REQUIRED_PACKAGES
    WHAT REQUIRED_PACKAGES_NOT_FOUND
)

if (NOT SAIL_BIN)
    message(FATAL_ERROR "Sail not found. See README.md for installation instructions.")
endif()
```

I guess we could add a `FindSail.cmake` file to turn it into a package but maybe that's overkill?

arichardson replied:
Probably easier to just keep the existing code.

arichardson commented on emulator/CMakeLists.txt:
```diff
@@ -0,0 +1,44 @@
+        PRIVATE softfloat sail_runtime GMP::GMP ZLIB::ZLIB
+    )
+
+    target_include_directories(emulator_${arch}
+        # So the generated C can find riscv_platform/prelude.h"
+        PRIVATE "${CMAKE_CURRENT_SOURCE_DIR}"
+    )
```
Another option would be to set [`CMAKE_INCLUDE_CURRENT_DIR`](https://cmake.org/cmake/help/latest/variable/CMAKE_INCLUDE_CURRENT_DIR.html) to add this globally

Timmmm replied:
Ah didn't know about that. Updated.

Timmmm replied:
Actually I couldn't get this to work for some reason, even if I put it in `emulator/CMakeLists.txt`.

arichardson replied:
Odd but explicit include dirs are probably better anyway.

arichardson commented on CMakeLists.txt:
```diff
@@ -0,0 +1,76 @@
+# Optional faster binary. Increases compilation time a lot though due to LTO.
+option(EXTRA_FAST "Enable aggressive optimisation flags (-march=native, -flto, etc)" FALSE)
+
+if (EXTRA_FAST)
+    include("cmake/extra_fast.cmake")
+endif()
```
I guess one problem with this is that by default CMake will give you a debug build so without instructions we might end up with people complaining that their LTO build is still super slow.

Might need to check that CMAKE_BUILD_TYPE is not empty and not `Debug`?

Timmmm replied:
Good point. I added a warning.

arichardson commented on cmake/extra_fast.cmake:
```diff
@@ -0,0 +1,28 @@
+
+# Try to use -march=x86-64-v3, but fall back to -march=native if we are
+# using an old compiler that doesn't support that flag.
+check_cxx_compiler_flag("-march=x86-64-v3" COMPILER_SUPPORTS_MARCH_X86_V3)
```
Not sure this is the best default to use, but it's probably fine since it will fail on e.g. AArch64 hosts and the fall back to -march=native.

jrtc27 replied:
This whole feature doesn’t belong here. It’s a separate change (and could equally have been done with make)

Timmmm replied:
Fair point, I'll defer it.

arichardson reviewed (approved):
A few minor comments but generally looks good to me.

Timmmm commented:
> if we do decide to keep it I have some suggestions so that you don’t have to modify their Makefile.

We aren't actually using their Makefile at all with the CMake version; I just kept the modifications so that it compiled as a standalone commit. I'll just remove the upgrade anyway though.

@arichardson I think you put a comment somewhere that I can't find now about keeping the Makefile. I think I'll just try and implement all of the features currently in the Makefile before merging this, rather than have two build systems...

arichardson reviewed (commented):
Adding all the other targets may take quite a while, maybe easiest for now to add CMake support and remove targets from the Makefile as we port them over?

E.g. this PR just removes the csim target (or rather have it defer to CMake `cmake -S . -B $(BUILD_DIR) -DCMAKE_BUILD_TYPE=Release && $(MAKE) -C $(BUILD_DIR)`) and then later commits add the extra features? Otherwise I worry this will be a long standing feature branch that never gets merged.

jrtc27 commented:
> Adding all the other targets may take quite a while, maybe easiest for now to add CMake support and remove targets from the Makefile as we port them over?
> 
> E.g. this PR just removes the csim target (or rather have it defer to CMake `cmake -S . -B $(BUILD_DIR) -DCMAKE_BUILD_TYPE=Release && $(MAKE) -C $(BUILD_DIR)`) and then later commits add the extra features? Otherwise I worry this will be a long standing feature branch that never gets merged.

Doing CMake incrementally makes sense to me, but I'd rather not incrementally remove things from Makefile. It's annoying to have to use both depending on what you want to do. I don't really see the benefit of it. Just leave Makefile alone until it's entirely subsumed by CMakeLists.txt and then delete it.

arichardson commented:
> > Adding all the other targets may take quite a while, maybe easiest for now to add CMake support and remove targets from the Makefile as we port them over?
> > E.g. this PR just removes the csim target (or rather have it defer to CMake `cmake -S . -B $(BUILD_DIR) -DCMAKE_BUILD_TYPE=Release && $(MAKE) -C $(BUILD_DIR)`) and then later commits add the extra features? Otherwise I worry this will be a long standing feature branch that never gets merged.
> 
> Doing CMake incrementally makes sense to me, but I'd rather not incrementally remove things from Makefile. It's annoying to have to use both depending on what you want to do. I don't really see the benefit of it. Just leave Makefile alone until it's entirely subsumed by CMakeLists.txt and then delete it.

That sounds even better to me

billmcspadden-riscv commented:
I agree.


On Thu, Jan 9, 2025 at 2:12 PM Alexander Richardson <
***@***.***> wrote:

> Adding all the other targets may take quite a while, maybe easiest for now
> to add CMake support and remove targets from the Makefile as we port them
> over?
> E.g. this PR just removes the csim target (or rather have it defer to
> CMake cmake -S . -B $(BUILD_DIR) -DCMAKE_BUILD_TYPE=Release && $(MAKE) -C
> $(BUILD_DIR)) and then later commits add the extra features? Otherwise I
> worry this will be a long standing feature branch that never gets merged.
>
> Doing CMake incrementally makes sense to me, but I'd rather not
> incrementally remove things from Makefile. It's annoying to have to use
> both depending on what you want to do. I don't really see the benefit of
> it. Just leave Makefile alone until it's entirely subsumed by
> CMakeLists.txt and then delete it.
>
> That sounds even better to me
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/riscv/sail-riscv/pull/647#issuecomment-2581163311>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXROLOBMH4VMSUCF5IGWAX32J3JZ3AVCNFSM6AAAAABT2RMBV2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKOBRGE3DGMZRGE>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Bill McSpadden
Formal Verification Engineer
RISC-V International
mobile:  503-807-9309

arichardson commented on .pre-commit-config.yaml:
```diff
@@ -1,6 +1,6 @@
 # See https://pre-commit.com for more information
 # See https://pre-commit.com/hooks.html for more hooks
-exclude: '^(prover_snapshots)|(generated_definitions)|(c_emulator/SoftFloat-3e)'
+exclude: '^(prover_snapshots)|(generated_definitions)|(dependencies/softfloat/berkeley-softfloat-3)'
```
I think we shouldn't run the checks on any dependencies? How about this:
```suggestion
exclude: '^(prover_snapshots)|(generated_definitions)|(dependencies)'
```

arichardson replied:
This is still open

Timmmm replied:
Oops, fixed.

Timmmm replied:
Oops sorry I forgot. It would still be nice to run pre-commit on the CMakeLists.txt file I added there, but I think something like this will work

```
exclude: '^(prover_snapshots)|(generated_definitions)|(dependencies/[^/]+/)'
```

arichardson replied:
Current version is good too, didn't not think about the cmakelists

arichardson commented on Makefile:
```diff
@@ -1,428 +0,0 @@
-# Select architecture: RV32 or RV64.
-ARCH ?= RV64
```
Need to keep makefile for now for the prover stuff?

arichardson commented on build_simulators.sh:
```diff
@@ -1,14 +0,0 @@
-}
-
-test_build make ARCH=RV32 c_emulator/riscv_sim_RV32
-test_build make ARCH=RV64 c_emulator/riscv_sim_RV64
```
Is this documented anywhere as a sensible script to use? If so we could just update it to use CMake?

arichardson commented on Makefile:
```diff
@@ -1,428 +0,0 @@
-
-# SAIL_FLAGS = -dtc_verbose 4
-
-ifneq (,$(COVERAGE))
-C_FLAGS += --coverage -O1
-SAIL_FLAGS += -Oconstant_fold
-else
-C_FLAGS += -O3 -flto=auto
-endif
-
-ifneq (,$(SAILCOV))
-ALL_BRANCHES = generated_definitions/c/all_branches
-C_FLAGS += -DSAILCOV
-SAIL_FLAGS += -c_coverage $(ALL_BRANCHES) -c_include sail_coverage.h
-C_LIBS += $(SAIL_LIB_DIR)/coverage/libsail_coverage.a -lm -lpthread -ldl
-endif
-
-# Optionally link C_LIBS statically. Unlike -static this will not
-# link glibc statically which is generally a bad idea.
-ifneq (,$(STATIC))
-    UNAME_S := $(shell sh -c 'uname -s 2>/dev/null || echo not')
-    ifeq ($(UNAME_S),Darwin)
-        # Unfortunately the Mac linker does not support -Bstatic.
-        GMP_LIBS = $(shell pkg-config --variable=libdir gmp)/libgmp.a
-        C_LIBS_WRAPPED = $(C_LIBS)
-    else
-        C_LIBS_WRAPPED = -Wl,--push-state -Wl,-Bstatic $(C_LIBS) -Wl,--pop-state
-    endif
-else
-    C_LIBS_WRAPPED = $(C_LIBS)
-endif
```
I believe we have lost support for these options. I've used the sailcov option before, but not the other ones.

arichardson commented on Makefile:
```diff
@@ -1,428 +0,0 @@
-endif
-	(cd generated_definitions/hol4/$(ARCH) && Holmake riscvTheory.uo)
-
-riscv_hol: generated_definitions/hol4/$(ARCH)/riscvScript.sml
-riscv_hol_build: generated_definitions/hol4/$(ARCH)/riscvTheory.uo
-.PHONY: riscv_hol riscv_hol_build
```
Is the hol target still working? If so do we need to port it over to cmake?

arichardson commented on c_emulator/CMakeLists.txt:
```diff
@@ -0,0 +1,54 @@
+    riscv_softfloat.h
+)
+
+foreach (arch IN ITEMS "rv32" "rv32d" "rv64" "rv64f" "rv32_rvfi" "rv32d_rvfi" "rv64_rvfi" "rv64f_rvfi")
```
Should we use variants here as well to make it easier to add the other targets in the future?

arichardson reviewed (commented):
Overall I still think this is a great change. A few things that were supported are no longer there.

I wonder if we should keep the Makefile for now and add something like the following to the top:
```make
ifndef I_REALLY_NEED_TO_USE_MAKE
$(error The Makefile build system is deprecated, please use CMake instead. If you are using a target that does not exist there yet, please file and issue and re-run make with I_REALLY_NEED_TO_USE_MAKE=1 until this has been fixed)
endif
```

jrtc27 commented on CMakeLists.txt:
```diff
@@ -0,0 +1,74 @@
+    message(FATAL_ERROR "Sail not found. See README.md for installation instructions.")
+endif()
+
+set(DEFAULT_ARCHITECTURES "rv32;rv64" CACHE STRING "Architectures to build by default (rv32|rv64|rv32d|rv64f)(_rvfi)? " )
```
It's not clear to me on first read what these things mean. My initial interpretation was:
rv32: no float
rv64: no float
rv32d: FLEN=64
rv64f: FLEN=32

But then that's odd because FLEN=64 is a sensible configuration for RV64 (more so than FLEN=32, even), and of course that's not actually what's going on, there's a default non-zero FLEN and then the suffixed ones are overrides. But those names are confusing and weird.

On top of that, this is a behavioural change. The current Makefile build always defaults to FLEN=64 even for RV32, but you've changed that default to be FLEN=32.

Timmmm replied:
You're right - I've changed it to `rv32f`, `rv32d`, `rv64f`, `rv64d` with the obvious meanings, and `rv32d` and `rv64d` are the defaults. I didn't add no-float configs since you can do that via a runtime option (the float length has to be compile-time currently).

arichardson commented:
Once this lands, I am happy to submit some follow-up improvements to support static builds and builds that build GMP from source as part of the build process to avoid any system dependencies.

jrtc27 commented:
> > I've rebased it, removed the `c_emulator` rename and the softfloat update (we can do that later), and added RVFI support. I also updated the CI scripts.
> > I'm not sure I'm much of a fan of having two build systems during a transition. Verilator has exactly this - an old `./configure && make` and an "experimental" CMake build system, and IMO it is quite confusing for users, and probably annoying for developers.
> > I reckon I can add the other targets without too much trouble, assuming they still work anyway! They aren't tested by CI and are presumably very rarely used.
> 
> I agree that having multiple build systems is annoying and confusing. What do you think of my suggestion to add a (opt-out) error to the makefile to point people towards CMake.
> 
> We should probably also update the README.

I mean, if the Makefile is needed in some cases then it shouldn't be an error to use it. And if it's not needed then it can be deleted.

It's not confusing at all to have both, *provided* the CMake one is clearly labelled as experimental and in-progress. What's confusing is when there are two documented ways of building (or one documented but the other one isn't clearly labelled as "don't use yet unless you really want to").

arichardson commented:
> > > I've rebased it, removed the `c_emulator` rename and the softfloat update (we can do that later), and added RVFI support. I also updated the CI scripts.
> > > I'm not sure I'm much of a fan of having two build systems during a transition. Verilator has exactly this - an old `./configure && make` and an "experimental" CMake build system, and IMO it is quite confusing for users, and probably annoying for developers.
> > > I reckon I can add the other targets without too much trouble, assuming they still work anyway! They aren't tested by CI and are presumably very rarely used.
> > 
> > 
> > I agree that having multiple build systems is annoying and confusing. What do you think of my suggestion to add a (opt-out) error to the makefile to point people towards CMake.
> > We should probably also update the README.
> 
> I mean, if the Makefile is needed in some cases then it shouldn't be an error to use it. And if it's not needed then it can be deleted.
> 
> It's not confusing at all to have both, _provided_ the CMake one is clearly labelled as experimental and in-progress. What's confusing is when there are two documented ways of building (or one documented but the other one isn't clearly labelled as "don't use yet unless you really want to").

I agree it shouldn't be an error when it's needed. I was just suggesting making it an error that you can override to find out if anyone is using the targets that don't exist in the CMkae world yet.

If we just add a deprecation warning or keep the makefile around unchanged no one is going to report it until it becomes an error.

Timmmm commented:
Ok I've done some updates:

1. Change float naming scheme as suggested by Jessica.
2. Add support for `-DCOVERAGE=TRUE` to enable Sail branch coverage support.
3. Change the `c_emulator/CMakelists.txt` variant list to use the same loop as in `model/CMakeLists.txt`. I really hope some of these compile time options will go away in future when Alasdair finishes the new config system.
4. Compile a few more variants in CI, to make sure they don't break.
5. I removed all of the whitelisted `riscv-tests` tests. We removed them because they don't work for some of the configs we use, but they do work for the default one so it makes sense to leave them for now I guess.
6. Added `build_simulators.sh` back for convenience.
7. Added the `Makefile` back and fixed an SMT argument.
9. Added support for rmem, and part of coq... but they're kind of incomplete because they don't actually compile at the moment - even with the original Makefile I run into various compilation errors, some in `.sail` files in the "standard library".
10. Add support for smt output. It compiles, but currently there are no `$property`s so it doesn't output anything.

Given rmem and coq are broken anyway I think maybe it is best to do what Alex/Jessica were suggesting and keep the Makefile but add a warning to use CMake instead.

arichardson commented on .github/workflows/compile.yml:
```diff
@@ -19,13 +19,24 @@ jobs:
+        # By default only the rv32d and rv64d emulators are build,
+        # but we want to build more targets here to ensure they
+        # can at least build without errors.
+        ninja c_emulator_rv32d c_emulator_rv64d c_emulator_rv32d_rvfi generated_smt_rv64f
```
IMO we should just build `ninja all` since that is what most people will be using.
If we add `EXCLUDE_FROM_ALL` to certain targets in CMake we can change that to be a sensible subset.

arichardson replied:
I see EXCLUDE_FROM_ALL is already handled - I'd suggest using `ninja all` + `ninja <additional targets>`

jrtc27 replied:
Also no need to cd here, use cmake -B and ninja -C.

Timmmm replied:
@arichardson I didn't use `ninja all` (or just `ninja` which is the same) because it only builds `rv32d` and `rv64d`. If we do `ninja && ninja <other targets>` then they won't run in parallel optimally. I can do `ninja all <other targets>` though; I think that makes the most sense.

@jrtc27 I had a go but I can't see how to get `ctest` to run without `cd` - you can do `ninja -C build test` but then you can't pass the junit argument.

jrtc27 replied:
ctest --test-dir?

jordancarlin replied:
That's strange. I just tried the following and it ran the tests successfully:
```
cmake -GNinja -DCMAKE_BUILD_TYPE=RelWithDebInfo -S . -B build
cmake --build build --target all riscv_sim_rv32d_rvfi generated_smt_rv64f generated_docs_rv64d
ctest --output-junit tests.xml --test-dir build
```

arichardson commented on CMakeLists.txt:
```diff
@@ -0,0 +1,73 @@
+# TODO: Add isabelle target.
+# TODO: Add lem target.
+# TODO: Add hol4 target.
+# TODO: Add latex target.
```
Not sure we still care about the latex output now that all the docs are asciidoc?

jrtc27 replied:
I assume ISAv10 will still need them to embed our own research extensions, and possibly the standard part too? Not sure if we've thought through how exactly ISAv10 will look in this world though.

arichardson replied:
Yeah but that target would exist in sail-cheri-riscv and not here? Although maybe by ISAv10 it would just be a fork of this repo?

Timmmm replied:
Yeah very good question. We should probably discuss it on Thursday. I'll remove the todo here anyway.

arichardson commented on build_simulators.sh:
```diff
@@ -1,14 +1,8 @@
+mkdir -p build
+cd build
+cmake .. -DCMAKE_BUILD_TYPE=Release
+make -j2 csim
```
why j2? Maybe something like `-j$(nproc || echo 2)`?

jrtc27 replied:
nproc isn't a thing on macOS (unless you install GNU coreutils)

(it does exist on FreeBSD these days though)

arichardson replied:
yeah that's why I added the fallback. I guess you could do something like this:
```suggestion
make -j$(nproc || sysctl -n hw.ncpu || echo 2) csim
```

jrtc27 replied:
That verbatim will shout at you over stderr though...

arichardson replied:
Yeah will need the right redirects. But then again I'm not sure this script really matters

Timmmm replied:
I picked 2 because it builds 2 simulators, and most of the time is just compiling the Sail, the `model.c` and linking which are all serial for each simulator.

I agree this script doesn't really matter though - it's probably more important as a kind of "how to compile" than something I'd expect people to actually use, in which case keeping it simple is probably more important than optimal CPU use? Happy to change it to e.g. 4 though.

arichardson replied:
Probably doesn't matter. Keeping two for the two simulators is fine IMO.

arichardson commented on c_emulator/CMakeLists.txt:
```diff
@@ -0,0 +1,67 @@
+            if (arch MATCHES "rvfi")
+                target_compile_definitions(c_emulator_${arch}
+                    PRIVATE RVFI_DII
+                )
```
Could use this if to skip rvfi targets from the default?
```suggestion
                )
                set_target_properties(c_emulator_${arch} PROPERTIES EXCLUDE_FROM_ALL)
```

arichardson replied:
Never mind, I see this is already handled above!

arichardson commented on sail_runtime/CMakeLists.txt:
```diff
@@ -0,0 +1,36 @@
+if (COVERAGE)
+    find_package(Threads REQUIRED)
+
+    # Note, you must run `make` in the ${sail_dir}/lib/coverage
```
Probably worth improving this in the future (e.g. building it automatically if not present), but not for this PR.

Timmmm replied:
Yes, definitely! See https://github.com/rems-project/sail/issues/893

arichardson commented on c_emulator/CMakeLists.txt:
```diff
@@ -0,0 +1,67 @@
+            # generates too many warnings to turn these on globally.
+
+            # target_compile_options(c_emulator_${arch} PRIVATE
+            #     -Wall -Wextra
```
IMO we should enable -Wall globally in the root CMakeLists.txt. Possibly even -Wextra although I imagine the softfloat library will not be happy, so that probably makes more sense to have just here.

Timmmm replied:
Yeah it's not just softfloat though (which isn't even happy with `-Wall`) - the `model.c` also generates a load of warnings. Once we've removed softfloat, and use the `#include "model.c"` trick for the model we can though.

arichardson commented on .github/workflows/compile.yml:
```diff
@@ -19,13 +19,24 @@ jobs:
+        # By default only the rv32d and rv64d emulators are build,
+        # but we want to build more targets here to ensure they
+        # can at least build without errors.
+        ninja c_emulator_rv32d c_emulator_rv64d c_emulator_rv32d_rvfi generated_smt_rv64f
```
```suggestion
        ninja all c_emulator_rv32d_rvfi generated_smt_rv64f
```

arichardson reviewed (approved):
I think this looks great now and since the Makefile is still there for now I don't see any reason to not merge it (once CI is happy).

Spotted a few minor things

jrtc27 commented on CMakeLists.txt:
```diff
@@ -0,0 +1,73 @@
+# Don't use `REQUIRED` so that we can print custom help messages.
+find_package(ZLIB)
+if (NOT ZLIB_FOUND)
+    message(FATAL_ERROR "Zlib not found. Try 'sudo apt install zlib1g-dev' or 'sudo dnf install zlib-devel'.")
```
I'd bet that an inexperienced user on macOS will turn up, hit one of these messages and try to run the commands given, then complain that it doesn't work...

Timmmm replied:
Fair point, and actually we can easily customise the message for the OS - I'll do that.

Timmmm replied:
This is done - I added `brew` messages for mac, and don't suggest anything for windows.

jrtc27 commented on CMakeLists.txt:
```diff
@@ -0,0 +1,73 @@
+
+find_package(GMP)
+if (NOT GMP_FOUND)
+    message(FATAL_ERROR "GMP not found. Try 'sudo apt install libgmp3-dev' or 'sudo dnf install gmp-devel'.")
```
```suggestion
    message(FATAL_ERROR "GMP not found. Try 'sudo apt install libgmp-dev' or 'sudo dnf install gmp-devel'.")
```

One should almost never depend on a specific library version in Debian, and in this case it's particularly pointless as libgmp3-dev is just a compatibility package that depends on the current major version one.

jrtc27 commented on .github/workflows/compile.yml:
```diff
@@ -19,13 +19,24 @@ jobs:
+        cd build
+        # Ninja is used because the CMake Makefile generator doesn't
+        # build top-level targets in parallel unfortunately.
+        cmake -GNinja -DCMAKE_BUILD_TYPE=Release ..
```
RelWithDebInfo is the correct choice for most use cases, not Release. But I also wonder if we should have the default be that, rather than Debug. Otherwise I fear people will just do `cmake && make`, end up with Debug builds and get the impression the model is slow when it's not (or at least not *that* slow), it's just user error.

Timmmm replied:
Hmm yeah. That is quite a problem in the Rust community too actually - people think it's slow because they did `cargo build` rather than `cargo build --release`.

I'm not sure we can change the default though with CMake? It's the default in `build_simulators.sh` at least though. (I'll change both to RelWithDebInfo).

jrtc27 replied:
```
if(NOT CMAKE_BUILD_TYPE)
  set(CMAKE_BUILD_TYPE RelWithDebInfo)
endif()
```
?

Timmmm replied:
Hmm yeah that makes me a bit nervous - CMake's build type is... weird in a few ways - there are multi-config build systems (MSVC and optionally Ninja) where it doesn't select a build type until build time (normally it's configure time) so you also need to check `CMAKE_CONFIGURATION_TYPES`.

I found [this code](https://www.kitware.com/cmake-and-the-default-build-type/) which probably works, but even there one of the comments suggests that the default empty-string build type was intended for distros that build everything from source and set custom flags (tbf I'm not sure we really care about that).

Anyway, do you mind if we defer this to a future PR? It might never be an issue if people don't make that mistake...

jrtc27 replied:
I mean, you've seen the kinds of questions that some people post on this repo...

Can we at least do what LLVM does and require you to explicitly set the build type?

https://github.com/llvm/llvm-project/blob/e6c9cd9c060c1fa8343398b9556a5a6c0f35d515/llvm/CMakeLists.txt#L89-L99

Timmmm replied:
Good idea, I've copied that. It looks like this:

```
❯ cmake ..
CMake Error at CMakeLists.txt:6 (message):
  

  No build type selected.  You need to pass -DCMAKE_BUILD_TYPE=<type> in
  order to configure the build.

      * -DCMAKE_BUILD_TYPE=Release        - Optimized build with no debug info.
      * -DCMAKE_BUILD_TYPE=RelWithDebInfo - Optimized build with debug info.
      * -DCMAKE_BUILD_TYPE=Debug          - Unoptimized build with debug info.
      * -DCMAKE_BUILD_TYPE=MinSizeRel     - Optimized for size instead of speed.


-- Configuring incomplete, errors occurred!

```

Timmmm commented:
Update:

* Per-OS error messages when GMP and Zlib aren't found.
* Update README
* Rename `Makefile` to `Makefile.old` to make it obvious at a glance that you shouldn't use the Makefile.

I think this is basically ready at this point. Probably any more improvements could be done in future PRs (and I'm starting to lose track of all the threads!)

jordancarlin commented:
The Makefile is still pointing at the old location for softfloat. I know we want to encourage everyone to use the new CMake flow, but if we are going to keep the old Makefile around I think we should make sure it is still functional. Or remove the c emulator targets from it and only leave the formal backends.

Timmmm commented:
Good point, I'll fix that. Also I think I forgot to do the RelWithDebInfo change - I'll do that in the morning.

jordancarlin commented:
My other concern is changing the name of the generated executable itself. Currently it's `riscv_sim_ARCH`, which makes it clear what it is if copied to a bin directory. The switch to `c_emulator_ARCH` removes this easy recognition and doesn't seem to provide much benefit. If we want to explicitly call out that it is the c emulator (as opposed to the SystemVerilog one) maybe something like `riscv_csim_ARCH` would be better.

Timmmm commented on Makefile.old:
```diff
@@ -1,4 +1,4 @@
-# WARNING! This Makefile is deprecated and has been *mostly* replaced by CMake.
+# WARNING! This Makefile.old is deprecated and has been *mostly* replaced by CMake.
```
Oops, fixed!

Timmmm commented:
I rebased it, updated it for the Coq output (which now builds with the latest `sail2` compiler, but it doesn't do everything the Makefile does), and also added a Github actions workflow to generate the JSON docs.

My aim for the latter is to run it manually and make a Github Release with it attached, so we can start including Sail snippets in the ISA manual ([as we do for CHERI](https://github.com/riscv/riscv-cheri/pull/498)).

jordancarlin commented on build_simulators.sh:
```diff
@@ -1,14 +1,8 @@
+mkdir -p build
+cd build
+cmake .. -DCMAKE_BUILD_TYPE=RelWithDebInfo
+make -j2 csim
```
Why not something like this instead? Primarily because it will respect the default generator if a user has overridden it using `$CMAKE_GENERATOR`. Also avoids the need for the `mkdir` and `cd`.
```
cmake -S . -B build -DCMAKE_BUILD_TYPE=RelWithDebInfo
cmake --build build -j2
```

Timmmm replied:
Done

jordancarlin replied:
`mkdir` shouldn’t be necessary when using the -B flag

arichardson commented on README.md:
```diff
@@ -47,53 +47,38 @@ Getting started
 
 Information on configuration options for the simulator is available from
-`./c_emulator/riscv_sim_<arch> -h`.
+`./riscv_sim_<arch> -h`.
```
Should this be prefixed with build/c_emulator?

arichardson commented on .github/workflows/compile.yml:
```diff
@@ -19,13 +19,24 @@ jobs:
+        # can at least build without errors.
+        ninja all riscv_sim_rv32d_rvfi generated_smt_rv64f generated_docs_rv64d
+        # These targets do not work with Sail 0.18: generated_rmem_rv32d_rmem generated_coq_rv64f_coq
+        ctest --output-junit text.xml
```
Does this actually work? I imagine it would need to be `tests.xml`?

arichardson replied:
Indeed it does not: `Warning: No files were found with the provided path: build/tests.xml. No artifacts will be uploaded.
`

arichardson commented on .github/workflows/compile.yml:
```diff
@@ -19,13 +19,24 @@ jobs:
       with:
         name: tests.xml
-        path: test/tests.xml
+        path: build/tests.xml
```
```suggestion
        path: build/tests.xml
        if-no-files-found: error
```
This would fix the problem above

Timmmm replied:
Why do people still design unsafe-by-default APIs? :facepalm:

arichardson commented on .github/workflows/compile.yml:
```diff
@@ -19,13 +19,24 @@ jobs:
     - name: Build and test simulators
-      run: test/run_tests.sh
+      run: |
+        mkdir build
+        cd build
+        # Ninja is used because the CMake Makefile generator doesn't
+        # build top-level targets in parallel unfortunately.
+        cmake -GNinja -DCMAKE_BUILD_TYPE=RelWithDebInfo ..
+        # By default only the rv32d and rv64d emulators are build,
+        # but we want to build more targets here to ensure they
+        # can at least build without errors.
+        ninja all riscv_sim_rv32d_rvfi generated_smt_rv64f generated_docs_rv64d
+        # These targets do not work with Sail 0.18: generated_rmem_rv32d_rmem generated_coq_rv64f_coq
+        ctest --output-junit text.xml
```
```suggestion
        # Ninja is used because the CMake Makefile generator doesn't
        # build top-level targets in parallel unfortunately.
        cmake -S . -B build -GNinja -DCMAKE_BUILD_TYPE=RelWithDebInfo
        # By default only the rv32d and rv64d emulators are build,
        # but we want to build more targets here to ensure they
        # can at least build without errors.
        ninja -C build all riscv_sim_rv32d_rvfi generated_smt_rv64f generated_docs_rv64d
        # These targets do not work with Sail 0.18: generated_rmem_rv32d_rmem generated_coq_rv64f_coq
        ctest --test-dir build --output-junit build/tests.xml
```
Would this work?

jordancarlin replied:
I was able to get it working and passing CI using this. The `build/tests.xml` in your version places the file in `build/build/tests.xml` because the `--test-dir` flag already gets us in the right directory. 

```
cmake -GNinja -DCMAKE_BUILD_TYPE=RelWithDebInfo -S . -B build
cmake --build build --target all riscv_sim_rv32d_rvfi generated_smt_rv64f generated_docs_rv64d
ctest --output-junit tests.xml --test-dir build
```

arichardson replied:
> I was able to get it working and passing CI using this. The `build/tests.xml` in your version places the file in `build/build/tests.xml` because the `--test-dir` flag already gets us in the right directory.
> 
> ```
> cmake -GNinja -DCMAKE_BUILD_TYPE=RelWithDebInfo -S . -B build
> cmake --build build --target all riscv_sim_rv32d_rvfi generated_smt_rv64f generated_docs_rv64d
> ctest --output-junit tests.xml --test-dir build
> ```

Ah thanks for testing this - I have updated my suggestion to drop the duplicate prefix.

Timmmm replied:
Ah nice work. I could have sworn I tried `--test-dir` and it didn't work... Must have done something stupid.

arichardson reviewed (approved):
Minor suggestion on the CI file but still LGTM

jordancarlin reviewed (approved):
Definitely more optimizations to make in the future, but at this point this seems good to go now that CI is properly functioning again.

Timmmm commented:
Ok I will merge this later today!

jrtc27 commented:
> ## Test Results
> 392 tests   - 4   392 ✅  - 4   49s ⏱️ +49s   1 suites  - 3     0 💤 ±0    1 files   ±0     0 ❌ ±0 
> 
> Results for commit [8818308](https://github.com/riscv/sail-riscv/commit/88183081b63b0de9d0eb9e7f7ddc739177ccac36). ± Comparison against base commit [8a75b29](https://github.com/riscv/sail-riscv/commit/8a75b297b116a1ffd8c62e98a7f43e2d93761d15).
> 
> This pull request removes 396 and adds 392 tests. Note that renamed tests count towards both.

Why have we lost 4 tests and 3 suites?

jrtc27 commented on .github/workflows/generate-json.yml:
```diff
@@ -0,0 +1,30 @@
+        cd build
+        cmake -GNinja -DCMAKE_BUILD_TYPE=Release ..
+        ninja generated_docs_rv64d
+    - name: Upload test results
```
test results?

Timmmm replied:
Oops, fixed.

jrtc27 commented on .github/workflows/generate-json.yml:
```diff
@@ -0,0 +1,30 @@
+        cmake -GNinja -DCMAKE_BUILD_TYPE=Release ..
+        ninja generated_docs_rv64d
+    - name: Upload test results
+      if: always()
```
I don't think that's what you want here?

Timmmm replied:
Good point, I removed that and added the `if-no-files-found: error` too.

jrtc27 commented on .github/workflows/compile.yml:
```diff
@@ -19,13 +19,23 @@ jobs:
         name: tests.xml
-        path: test/tests.xml
+        path: build/tests.xml
+        if-no-files-found: error
```
With the if: always() this will give cascading errors, but that's probably ok. Really what we want is "if we ran the test suite then upload test results regardless of exit code for ctest, but still fail the build if it had failures".

Timmmm replied:
Yeah I mainly use Gitlab CI so I'm not really sure how to do that, sorry! I guess 2 errors when you expect 1 is not too bad.

Timmmm commented:
> Why have we lost 4 tests and 3 suites?

Good question - the 4 lost are because the old output had compiling the models as tests.

```
Building 32-bit RISCV C emulator
Building 32-bit RISCV RVFI C emulator
Building 64-bit RISCV C emulator
Building 64-bit RISCV RVFI C emulator
```

The 3 suites are because the previous test runner script split the tests up into suites based on the above 4 models. In the new version they are combined.

I double checked that it is running all the same tests. You can see the list on the build CI job.

Note, the old script compiles the RVFI emulators but doesn't actually run any tests on them. This CMake version does the same.
